### PR TITLE
Follow-up incremental joins

### PR DESCRIPTION
**Pull request by kim**
# Description of Changes

Adds commentary and rejects unsupported subscription queries early (instead of never returning any updates).

# API and ABI

 - [ ] This is a breaking change to the module ABI
 - [ ] This is a breaking change to the module API
 - [ ] This is a breaking change to the ClientAPI
 - [ ] This is a breaking change to the SDK API

*If the API is breaking, please state below what will break*

# Expected complexity level and risk

*How complicated do you think these changes are? Grade on a scale from 1 to 5,
where 1 is a trivial change, and 5 is a deep-reaching and complex change.*

*This complexity rating applies not only to the complexity apparent in the diff,
but also to its interactions with existing and future code.*

*If you answered more than a 2, explain what is complex about the PR,
and what other components it interacts with in potentially concerning ways.*

### CODE REVIEW

File: crates/core/src/subscription/query.rs [hunk 1]
```diff
@@ -866,27 +886,51 @@ mod tests {
...
-
-        for q in &q {
-            assert_eq!(
-                run_query(&db, &mut tx, q, AuthCtx::for_testing())?.len(),
-                1,
-                "Not return results"
-            );
+        run(&db, &mut tx, sql_insert, AuthCtx::for_testing())?;
+
+        let sql_query = "\
```
Comment (kim) [inline crates/core/src/subscription/query.rs:922]: This is not detected as an IndexJoin, despite actually creating the indexes above. It looks to be due to the first `query` element being an `IndexScan`, and the length thus being > 2. Not sure how to fix.

File: crates/core/src/host/wasm_common/module_host_actor.rs [hunk 2]
```diff
@@ -258,11 +258,18 @@ impl<T: WasmModule> Module for WasmModuleHostActor<T> {
...
-            let compiled = crate::subscription::query::compile_read_only_query(db, tx, &auth, &query)?;
 
-            sql::execute::execute_sql(db, tx, compiled.into_iter().map(CrudExpr::Query).collect(), auth)
+            let compiled = sql::compiler::compile_sql(db, tx, &query)?
+                .into_iter()
+                .map(|expr| {
+                    if matches!(expr, CrudExpr::Query { .. }) {
+                        Ok(expr)
+                    } else {
+                        Err(anyhow!("One-off queries are not allowed to modify the database"))
```
Comment (kim) [inline crates/core/src/host/wasm_common/module_host_actor.rs:268]: @kazimuth Could you eyeball this? I'm wondering if we actually want to return an error here -- we could also just filter out any non-`SELECT` statements and return empty results for them.

File: crates/core/src/subscription/query.rs (hunk 1 above)
Comment (kim) [reply crates/core/src/subscription/query.rs:922]: Well yes, I thought about that. But the problem is something else. This is how the compiled query (with qualified wildcard) looks like; as you can see, `query` contains more than two elements, and the first one is `IndexScan`:

<details>
  <summary>Click to expand</summary>

```rust
QueryExpr {
    source: DbTable(
        DbTable {
            head: Header {
                table_name: "MobileEntityState",
                fields: [
                    Column {
                        field: Name {
                            table: "MobileEntityState",
                            field: "entity_id",
                        },
                        algebraic_type: Builtin(
                            U64,
                        ),
                    },
                    Column {
                        field: Name {
                            table: "MobileEntityState",
                            field: "location_x",
                        },
                        algebraic_type: Builtin(
                            I32,
                        ),
                    },
                    Column {
                        field: Name {
                            table: "MobileEntityState",
                            field: "location_z",
                        },
                        algebraic_type: Builtin(
                            I32,
                        ),
                    },
                    Column {
                        field: Name {
                            table: "MobileEntityState",
                            field: "destination_x",
                        },
                        algebraic_type: Builtin(
                            I32,
                        ),
                    },
                    Column {
                        field: Name {
                            table: "MobileEntityState",
                            field: "destination_z",
                        },
                        algebraic_type: Builtin(
                            I32,
                        ),
                    },
                    Column {
                        field: Name {
                            table: "MobileEntityState",
                            field: "is_running",
                        },
                        algebraic_type: Builtin(
                            Bool,
                        ),
                    },
                    Column {
                        field: Name {
                            table: "MobileEntityState",
                            field: "timestamp",
                        },
                        algebraic_type: Builtin(
                            U64,
                        ),
                    },
                    Column {
                        field: Name {
                            table: "MobileEntityState",
                            field: "dimension",
                        },
                        algebraic_type: Builtin(
                            U32,
                        ),
                    },
                ],
            },
            table_id: 6,
            table_type: User,
            table_access: Public,
        },
    ),
    query: [
        IndexScan(
            IndexScan {
                table: DbTable {
                    head: Header {
                        table_name: "MobileEntityState",
                        fields: [
                            Column {
                                field: Name {
                                    table: "MobileEntityState",
                                    field: "entity_id",
                                },
                                algebraic_type: Builtin(
                                    U64,
                                ),
                            },
                            Column {
                                field: Name {
                                    table: "MobileEntityState",
                                    field: "location_x",
                                },
                                algebraic_type: Builtin(
                                    I32,
                                ),
                            },
                            Column {
                                field: Name {
                                    table: "MobileEntityState",
                                    field: "location_z",
                                },
                                algebraic_type: Builtin(
                                    I32,
                                ),
                            },
                            Column {
                                field: Name {
                                    table: "MobileEntityState",
                                    field: "destination_x",
                                },
                                algebraic_type: Builtin(
                                    I32,
                                ),
                            },
                            Column {
                                field: Name {
                                    table: "MobileEntityState",
                                    field: "destination_z",
                                },
                                algebraic_type: Builtin(
                                    I32,
                                ),
                            },
                            Column {
                                field: Name {
                                    table: "MobileEntityState",
                                    field: "is_running",
                                },
                                algebraic_type: Builtin(
                                    Bool,
                                ),
                            },
                            Column {
                                field: Name {
                                    table: "MobileEntityState",
                                    field: "timestamp",
                                },
                                algebraic_type: Builtin(
                                    U64,
                                ),
                            },
                            Column {
                                field: Name {
                                    table: "MobileEntityState",
                                    field: "dimension",
                                },
                                algebraic_type: Builtin(
                                    U32,
                                ),
                            },
                        ],
                    },
                    table_id: 6,
                    table_type: User,
                    table_access: Public,
                },
                col_id: 1,
                lower_bound: Excluded(
                    I32(
                        96000,
                    ),
                ),
                upper_bound: Excluded(
                    I32(
                        192000,
                    ),
                ),
            },
        ),
        Select(
            Cmp {
                op: Logic(
                    And,
                ),
                lhs: Cmp {
                    op: Cmp(
                        Gt,
                    ),
                    lhs: Field(
                        Name(
                            Name {
                                table: "MobileEntityState",
                                field: "location_z",
                            },
                        ),
                    ),
                    rhs: Field(
                        Value(
                            I32(
                                96000,
                            ),
                        ),
                    ),
                },
                rhs: Cmp {
                    op: Cmp(
                        Lt,
                    ),
                    lhs: Field(
                        Name(
                            Name {
                                table: "MobileEntityState",
                                field: "location_z",
                            },
                        ),
                    ),
                    rhs: Field(
                        Value(
                            I32(
                                192000,
                            ),
                        ),
                    ),
                },
            },
        ),
        JoinInner(
            JoinExpr {
                rhs: QueryExpr {
                    source: DbTable(
                        DbTable {
                            head: Header {
                                table_name: "EnemyState",
                                fields: [
                                    Column {
                                        field: Name {
                                            table: "EnemyState",
                                            field: "entity_id",
                                        },
                                        algebraic_type: Builtin(
                                            U64,
                                        ),
                                    },
                                    Column {
                                        field: Name {
                                            table: "EnemyState",
                                            field: "herd_id",
                                        },
                                        algebraic_type: Builtin(
                                            I32,
                                        ),
                                    },
                                    Column {
                                        field: Name {
                                            table: "EnemyState",
                                            field: "status",
                                        },
                                        algebraic_type: Builtin(
                                            I32,
                                        ),
                                    },
                                    Column {
                                        field: Name {
                                            table: "EnemyState",
                                            field: "type",
                                        },
                                        algebraic_type: Builtin(
                                            I32,
                                        ),
                                    },
                                    Column {
                                        field: Name {
                                            table: "EnemyState",
                                            field: "direction",
                                        },
                                        algebraic_type: Builtin(
                                            I32,
                                        ),
                                    },
                                ],
                            },
                            table_id: 7,
                            table_type: User,
                            table_access: Public,
                        },
                    ),
                    query: [],
                },
                col_lhs: Name {
                    table: "MobileEntityState",
                    field: "entity_id",
                },
                col_rhs: Name {
                    table: "EnemyState",
                    field: "entity_id",
                },
            },
        ),
        Project(
            [
                Name(
                    Name {
                        table: "MobileEntityState",
                        field: "entity_id",
                    },
                ),
                Name(
                    Name {
                        table: "MobileEntityState",
                        field: "location_x",
                    },
                ),
                Name(
                    Name {
                        table: "MobileEntityState",
                        field: "location_z",
                    },
                ),
                Name(
                    Name {
                        table: "MobileEntityState",
                        field: "destination_x",
                    },
                ),
                Name(
                    Name {
                        table: "MobileEntityState",
                        field: "destination_z",
                    },
                ),
                Name(
                    Name {
                        table: "MobileEntityState",
                        field: "is_running",
                    },
                ),
                Name(
                    Name {
                        table: "MobileEntityState",
                        field: "timestamp",
                    },
                ),
                Name(
                    Name {
                        table: "MobileEntityState",
                        field: "dimension",
                    },
                ),
            ],
            Some(
                6,
            ),
        ),
    ],
}
```

</details>

File: crates/core/src/subscription/query.rs [hunk 3]
```diff
@@ -119,18 +123,28 @@ pub fn compile_read_only_query(
...
 }
 
-#[derive(Clone, Copy, Debug, Eq, PartialEq)]
+/// The kind of [`QueryExpr`] currently supported for incremental evaluation.
+#[derive(Clone, Copy, Debug, Eq, Ord, PartialEq, PartialOrd)]
 pub enum Supported {
+    /// A scan or [`QueryExpr::Select`] of a single table.
+    ///
+    /// Note that non-wildcard projections are accepted, but effectively ignored
+    /// (a matching row is always returned in its entirety).
```
[the remark this answers is not in the document]
Comment (kim) [reply crates/core/src/subscription/query.rs:138]: Maybe I dreamt that up. But then, we should reject those, no? Clients will not be able to deal with them afaict.

[the remark this answers is not in the document]
Comment (kim) [reply crates/core/src/subscription/query.rs]: Well, nevermind. I removed that comment.

**Review by joshua-spacetime — APPROVED** [review]

**Pull request merged**